### DISCUSSION TITLE
canceling conflicting previous ongoing mutations

### DISCUSSION THREAD

**Discussion by villesau** [urql-graphql/urql#2264] [Q&A]
Hi,

is it possible to cancel ongoing mutations when conflicting mutation is fired? I'm having `like` & `removeLike` mutations. Both are optimistic so the result reflects on the UI instantaneously. However, when clicking the button multiple times in a short time, the logic becomes flaky and the end result might be something unexpected due to network delays. Button might indicate liked, but when the final request returns, it might update to be non-liked. An alternative way would be to queue the conflicting mutations so that the order is always deterministic.

E: Looks like similar happens even when consolidating the mutation to `like` only which takes boolean as an argument.

**Comment by Leksat** [comment]
Even if a mutation is cancelled, it does not mean it haven't reached the server. So the end result might be even more unexpected 😄 

Instead we can queue mutations. And optionally limit the queue size so that we don't fire unnecessary requests. Something like this:
```ts
import PQueue from 'p-queue';
import { useState } from 'react';
import { useMutation } from 'urql';

function Widget() {
  const [, doMutate] = useMutation(`mutation MyMutation { ... }`);
  const [queue] = useState(new PQueue({ concurrency: 1 }));
  return (
    <button
      onClick={() => {
        if (queue.size > 0) {
          queue.clear();
        }
        queue.add(async () => doMutate({}));
      }}
    >
      Mutate
    </button>
  );
}
```
